File: src/api/generate.ts
// API bindings for Tauri commands
import { invoke } from '@tauri-apps/api/core';
import type { AIModel, ImageSize } from '../types';

export interface GenerateRequest {
    model: string;
    prompt: string;
    image_base64: string;
    mask_base64: string;
    reference_images?: string[]; // Optional reference images as base64
    image_size?: ImageSize;
}

export interface GenerateResponse {
    success: boolean;
    image_base64: string | null;
    error: string | null;
}

export interface CompositeRequest {
    base_image_base64: string;
    patch_image_base64: string;
    x: number;
    y: number;
    target_width: number;
    target_height: number;
    format: string;
}

export interface CompositeResponse {
    success: boolean;
    image_base64: string | null;
    error: string | null;
}

/**
 * Generate fill for a selected region using Nano Banana API
 * @param model - Which model to use
 * @param prompt - Text description of what to generate
 * @param imageBase64 - The cropped source image as base64
 * @param maskBase64 - The mask image as base64
 * @param referenceImages - Optional reference images to guide generation
 */
export async function generateFill(
    model: AIModel,
    prompt: string,
    imageBase64: string,
    maskBase64: string,
    referenceImages: string[] = [],
    imageSize?: ImageSize
): Promise<GenerateResponse> {
    const request: GenerateRequest = {
        model,
        prompt,
        image_base64: imageBase64,
        mask_base64: maskBase64,
        reference_images: referenceImages,
        image_size: imageSize,
    };

    return invoke<GenerateResponse>('generate_fill', { request });
}

/**
 * Composite a generated patch back onto the base image
 */
export async function compositePatch(
    baseImageBase64: string,
    patchImageBase64: string,
    x: number,
    y: number,
    targetWidth: number,
    targetHeight: number,
    format: string = 'png'
): Promise<CompositeResponse> {
    const request: CompositeRequest = {
        base_image_base64: baseImageBase64,
        patch_image_base64: patchImageBase64,
        x,
        y,
        target_width: targetWidth,
        target_height: targetHeight,
        format,
    };

    return invoke<CompositeResponse>('composite_patch', { request });
}

// === Layer Compositing ===

export interface LayerData {
    id: string;
    image_data: string;
    visible: boolean;
    opacity: number; // 0-100
    x?: number;
    y?: number;
    width?: number;
    height?: number;
    blend_mode?: string;
}

export interface CompositeLayersRequest {
    layers: LayerData[];
    canvas_width: number;
    canvas_height: number;
    format: string;
}

export interface CompositeLayersResponse {
    success: boolean;
    image_base64: string | null;
    error: string | null;
}

/**
 * Composite all visible layers into a single image
 */
export async function compositeLayers(
    layers: LayerData[],
    canvasWidth: number,
    canvasHeight: number,
    format: string = 'png'
): Promise<CompositeLayersResponse> {
    const request: CompositeLayersRequest = {
        layers,
        canvas_width: canvasWidth,
        canvas_height: canvasHeight,
        format,
    };

    return invoke<CompositeLayersResponse>('composite_layers', { request });
}

/**
 * Store the API key securely
 */
export async function setApiKey(apiKey: string): Promise<void> {
    return invoke('set_api_key', { apiKey });
}

/**
 * Check if API key is configured
 */
export async function hasApiKey(): Promise<boolean> {
    return invoke<boolean>('has_api_key');
}

/**
 * Delete the stored API key
 */
export async function deleteApiKey(): Promise<void> {
    return invoke('delete_api_key');
}

File: src-tauri/src/commands/generate.rs
// BananaSlice - Generation Commands
// Tauri commands for AI image generation

use crate::api::{Model, NanoBananaClient};
use crate::keystore;
use base64::{engine::general_purpose::STANDARD, Engine};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Serialize, Deserialize)]
pub struct GenerateRequest {
    pub model: String,
    pub prompt: String,
    pub image_base64: String,
    pub mask_base64: String,
    #[serde(default)]
    pub reference_images: Vec<String>, // Optional reference images as base64
    #[serde(default)]
    pub image_size: Option<String>, // Optional output resolution: 1K, 2K, 4K
}

#[derive(Debug, Serialize)]
pub struct GenerateResponse {
    pub success: bool,
    pub image_base64: Option<String>,
    pub error: Option<String>,
}

/// Get debug output directory
fn get_debug_dir() -> PathBuf {
    let path = PathBuf::from("C:/Users/sohan/Desktop/Projects/BananaSlice/debug_output");
    if !path.exists() {
        let _ = fs::create_dir_all(&path);
    }
    path
}

/// Save base64 image to file for debugging
fn save_debug_image(base64_data: &str, filename: &str) {
    if let Ok(bytes) = STANDARD.decode(base64_data) {
        let path = get_debug_dir().join(filename);
        if let Err(e) = fs::write(&path, bytes) {
            log::error!("Failed to save debug image {}: {}", filename, e);
        } else {
            log::info!("Saved debug image: {:?}", path);
        }
    }
}

/// Generate fill for a selected region
#[tauri::command]
pub async fn generate_fill(request: GenerateRequest) -> GenerateResponse {
    // Save input images for debugging
    log::info!("=== DEBUG: Saving input images ===");
    save_debug_image(&request.image_base64, "01_input_cropped.png");
    save_debug_image(&request.mask_base64, "02_input_mask.png");
    
    // Get API key from secure storage
    let api_key = match keystore::get_api_key() {
        Ok(key) => key,
        Err(_) => {
            return GenerateResponse {
                success: false,
                image_base64: None,
                error: Some("API key not configured. Please set your Gemini API key in Settings.".to_string()),
            };
        }
    };

    // Parse model
    let model = match request.model.as_str() {
        "nano-banana-pro" => Model::NanoBananaPro,
        "nano-banana" | _ => Model::NanoBanana,
    };

    // Create client and make request
    let client = NanoBananaClient::new(api_key);
    
    // Convert reference images to &str slices
    let ref_images: Vec<&str> = request.reference_images.iter().map(|s| s.as_str()).collect();
    
    match client
        .generate_fill(
            model,
            &request.prompt,
            &request.image_base64,
            &request.mask_base64,
            &ref_images,
            request.image_size.as_deref(),
        )
        .await
    {
        Ok(image_base64) => {
            // Save output image for debugging
            log::info!("=== DEBUG: Saving output image ===");
            save_debug_image(&image_base64, "03_output_generated.png");
            
            GenerateResponse {
                success: true,
                image_base64: Some(image_base64),
                error: None,
            }
        },
        Err(e) => GenerateResponse {
            success: false,
            image_base64: None,
            error: Some(e.to_string()),
        },
    }
}

/// Store the API key securely
#[tauri::command]
pub fn set_api_key(api_key: String) -> Result<(), String> {
    keystore::store_api_key(&api_key).map_err(|e| e.to_string())
}

/// Check if API key is configured
#[tauri::command]
pub fn has_api_key() -> bool {
    keystore::has_api_key()
}

/// Delete the stored API key
#[tauri::command]
pub fn delete_api_key() -> Result<(), String> {
    keystore::delete_api_key().map_err(|e| e.to_string())
}
